### PR TITLE
Surface gradient decal blending

### PR DESCRIPTION
**Pull request by adrien-de-tocqueville**
### Purpose of this PR

This PR adds a new optional way for HDRP to blend normal maps for decals: it uses surface gradient to apply decal normal while preserving the underlying object normal.

Before:
surf-master

After:
surf-branch

This is enabled via an option in the HDRP Asset.
It has an other option to enable high precision buffer: If disabled, the normal will have max angle of 45°:

image

---
### Testing status

Tested with HDRP/Decal and ShaderGraph Decal on all shader types: Lit (layered/tessellated), AxF, Terrain, ShaderGraph
Tested that disabling the option gives the same result as before

Added graphic test
1714_DecalSurfaceGradient

---
### Comments to reviewers

This introduce a change in the way decal data is stored in the DBuffer, which will break backward compatibility with existing custom decal shaders. We should mention it the the upgrade guide and try to output a compilation error for shaders using the old technique

### CODE REVIEW

**Review by JMargevics — APPROVED** [review]
enabled
switch

Feature looks good to me. Things I've tried: 

- Upgrading decals from current master
- Enabling HDRP asset option
- With or without high precision buffer
- Checked tooltips
- Inside LookDev
- Undo/Redo

@adrien-de-tocqueville said he works on a test to cover the automation. 

All good ✅

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 1]
```diff
@@ -1088,7 +1088,7 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
 #if ENABLE_NVIDIA && ENABLE_NVIDIA_MODULE
             m_DebugDisplaySettings.nvidiaDebugView.Update();
 #endif
-            Terrain.GetActiveTerrains(m_ActiveTerrains);
+            //Terrain.GetActiveTerrains(m_ActiveTerrains);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1091]: is it expected? I dot' think so

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 2]
```diff
@@ -1463,8 +1463,8 @@ ref renderDatas
                         settingsCopy.forceResolution = true;
                         DynamicResolutionHandler.UpdateAndUseCamera(camera, settingsCopy);
 
-                        foreach (var terrain in m_ActiveTerrains)
-                            terrain.SetKeepUnusedCameraRenderingResources(camera.GetInstanceID(), true);
+                        //foreach (var terrain in m_ActiveTerrains)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:1466]: is it expected? I dot' think so

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Settings/RenderPipelineSettings.cs [hunk 3]
```diff
@@ -94,6 +94,8 @@ internal static RenderPipelineSettings NewDefault()
                 supportedLitShaderMode = SupportedLitShaderMode.DeferredOnly,
                 supportDecals = true,
                 supportDecalLayers = false,
+                supportSurfaceGradient = false,
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Settings/RenderPipelineSettings.cs:97]: should be true by default. We want it for new project but keep it false for old project (can be done wit the migration step)

**Review by sebastienlagarde — COMMENTED** [review]
You forget to strip the extra variant in the stripper

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**